Providing option to calculate exact objective

walragatver opened the pull request:
Hi, this is for #108. It looks like we need to modify some `tests` in the `mlpack` repository after this change. I will start working on it once this gets approved.

zoq commented on include/ensmallen_bits/ada_delta/ada_delta.hpp:
```diff
@@ -64,6 +64,9 @@ class AdaDelta
    * @param tolerance Maximum absolute tolerance to terminate algorithm.
    * @param shuffle If true, the function order is shuffled; otherwise, each
    *        function is visited in linear order.
+   * @param exactObjective Flag that determines whether actual objective over
```
Sounds a little bit vague to me, what about "Calculate the exact objective (Default: estimate the final objective obtained on the last pass over the data)."

walragatver replied:
Yes, this is quite clear.

zoq commented on include/ensmallen_bits/ada_delta/ada_delta.hpp:
```diff
@@ -128,6 +132,11 @@ class AdaDelta
   //! Modify whether or not the individual functions are shuffled.
   bool& Shuffle() { return optimizer.Shuffle(); }
 
+  //! Get whether or not the actual objective is calculated after training.
```
Not sure, I would probably remove the "after training" part, see this more as an optimisation process.

walragatver replied:
Sorry, I just got carried away.

rcurtin commented on include/ensmallen_bits/cne/cne_impl.hpp:
```diff
@@ -119,7 +121,11 @@ double CNE::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
+  {
+    lastBestFitness = function.Evaluate(iterate);
+  }
+  return lastBestFitness;
```
Since CNE uses only an arbitrary function type, there's no need for `exactObjective` here.  It only applies to optimizers that optimize decomposable functions.  I think the documentation is a little wrong for CNE though (I updated it in another PR that I'm about to open).

rcurtin commented on include/ensmallen_bits/scd/scd_impl.hpp:
```diff
@@ -89,8 +91,12 @@ double SCD<DescentPolicyType>::Optimize(ResolvableFunctionType& function,
+  {
+    lastObjective = function.Evaluate(iterate);
+  }
+  return lastObjective;
```
Just like CNE, SCD doesn't work on decomposable functions, so I think we can leave SCD as it is. :+1:

rcurtin commented on include/ensmallen_bits/spsa/spsa_impl.hpp:
```diff
@@ -98,8 +100,12 @@ inline double SPSA::Optimize(
+  {
+    overallObjective = function.Evaluate(iterate);
+  }
+  return overallObjective;
```
SPSA doesn't work on decomposable functions, so we can avoid the change here too.

walragatver commented:
Hi @rcurtin , @zoq ,
This is now ready to merge. I have currently kept tommorow's date in `History.md`.

zoq commented on HISTORY.md:
```diff
@@ -1,3 +1,7 @@
+### ensmallen 1.1?.?
+###### ????-??-??
+  * Add option to avoid computing exact objective at end of optimization (#109).
```
Super picky comment: "at the end of the optimization".

zoq commented on doc/optimizers.md:
```diff
@@ -1485,11 +1497,12 @@ the following other optimizers:
  The attributes of the optimizer may also be modified via the member methods
  `StepSize()`, `BatchSize()`, `Beta1()`, `Beta2()`, `Eps()`, `MaxIterations()`,
- `Tolerance()`, `Shuffle()`,`V1()`,`V2()`, and `ResetPolicy()`.
+ `Tolerance()`, `Shuffle()`, `ExactObjective()`, `V1()`, `V2()`, and `ResetPolicy()`.
```
Nice catch!

walragatver commented:
Hi @zoq,
Thanks for the final approval. I have made the required changes.

coatless commented:
Regarding the parameter ordering, the current function call is:

```
AdaDelta(stepSize, batchSize, rho, epsilon, maxIterations, tolerance, shuffle, resetPolicy)
```

In this update, the function call switches to: 

```
AdaDelta(stepSize, batchSize, rho, epsilon, maxIterations, tolerance, shuffle, **exactObjective**, resetPolicy)
```

Perhaps the addition of `exactObjective` should be at the end of the function definition?

```
AdaDelta(stepSize, batchSize, rho, epsilon, maxIterations, tolerance, shuffle, resetPolicy, **exactObjective**)
```

walragatver commented:
@coatless, Thanks for raising the issue. I also wanted to discuss that. If you look in some of the cases `UpdateRule`  parameter comes after the `shuffle` parameter. If we keep `exactObjective` after that then in order to use it a user will need to provide the `updateRule` parameter always. I thought it's not a great idea. So I kept `exactObjective` after `shuffle` as both are `bools` indicating some optional functionality. 
Let me know if this is incorrect.

zoq commented:
Good point, I think I rather change the update policies instead of changing the behavior of the objective calculation. So personally I would add the value at the end. Note this will be part of the next major release so we don't necessarily have to worry about backward compatibility, but I think it would be great to have anyway.

zoq commented:
Let us keep this open, I think this is close to being merged.

rcurtin commented:
It seems to me like the only thing that remains is to fix the argument ordering issue.  I agree with @zoq on what he wrote:

> Good point, I think I rather change the update policies instead of changing the behavior of the objective calculation. So personally I would add the value at the end.

walragatver commented:
Hi @rcurtin, 
 Sorry for the slow response. Instead of adding a new constructor I have added `exactObjective` to the last position. I did that because it was difficult to `overload `the constructor because the last three arguments are of same `type`(`bool`). I have modified everything and this is ready to merge.

rcurtin commented:
@walragatver awesome work, thanks again! :+1:
